// gettext-merger/lib/cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var src_exports = {};
__export(src_exports, {
  Block: () => import_block.Block,
  SetOfBlocks: () => import_setOfBlocks.SetOfBlocks,
  extractPotHeader: () => import_utils.extractPotHeader,
  mergeBlocks: () => mergeBlocks,
  mergeComments: () => mergeComments,
  mergePotFile: () => mergePotFile,
  mergePotFileContent: () => mergePotFileContent,
  mergePotObject: () => mergePotObject
});
module.exports = __toCommonJS(src_exports);
var import_promises = __toESM(require("fs/promises"));
var import_setOfBlocks = require("./setOfBlocks.js");
var import_fs = require("./fs.js");
var import_block = require("./block.js");
var import_utils = require("./utils");
function mergeBlocks(...arrays) {
  const set = new import_setOfBlocks.SetOfBlocks();
  for (const array of arrays) {
    set.addArray(array);
  }
  return set;
}
__name(mergeBlocks, "mergeBlocks");
function mergePotFile(filePaths) {
  return __async(this, null, function* () {
    const headers = [];
    const mergedSet = yield Promise.all(
      filePaths.map((filePath) => __async(this, null, function* () {
        const fileContent = yield import_promises.default.readFile(filePath, "utf8");
        const [header, content] = (0, import_utils.extractPotHeader)(fileContent);
        if (header)
          headers.push(header);
        return new import_setOfBlocks.SetOfBlocks((0, import_fs.parseFileContent)(content)).blocks;
      }))
    );
    const currentBlocks = Array.from(mergedSet);
    return [Array.from(headers), mergeBlocks(...currentBlocks)];
  });
}
__name(mergePotFile, "mergePotFile");
function mergePotFileContent(fileContents) {
  const mergedSet = fileContents.map((content) => {
    return new import_setOfBlocks.SetOfBlocks((0, import_fs.parseFileContent)(content)).blocks;
  });
  const currentBlocks = Array.from(mergedSet);
  return mergeBlocks(...currentBlocks).toStr();
}
__name(mergePotFileContent, "mergePotFileContent");
function mergePotObject(translationObject) {
  const mergedSet = translationObject.map((content) => {
    return new import_setOfBlocks.SetOfBlocks(content.blocks).blocks;
  });
  return mergeBlocks(...mergedSet);
}
__name(mergePotObject, "mergePotObject");
function mergeUnique(current = [], other = []) {
  const mergeSet = /* @__PURE__ */ new Set([...current, ...other]);
  return Array.from(mergeSet);
}
__name(mergeUnique, "mergeUnique");
function mergeComments(current, other) {
  return {
    translator: mergeUnique(current == null ? void 0 : current.translator, other == null ? void 0 : other.translator),
    reference: mergeUnique(current == null ? void 0 : current.reference, other == null ? void 0 : other.reference),
    extracted: mergeUnique(current == null ? void 0 : current.extracted, other == null ? void 0 : other.extracted),
    flag: mergeUnique(current == null ? void 0 : current.flag, other == null ? void 0 : other.flag).join("\n"),
    previous: mergeUnique(current == null ? void 0 : current.previous, other == null ? void 0 : other.previous)
  };
}
__name(mergeComments, "mergeComments");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Block,
  SetOfBlocks,
  extractPotHeader,
  mergeBlocks,
  mergeComments,
  mergePotFile,
  mergePotFileContent,
  mergePotObject
});
